Sources/Sentry/SentryOptions.m

```diff
@@ -14,60 +14,86 @@ @implementation SentryOptions
     ];
 }
 
+- (instancetype)init
+{
+    if (self = [super init]) {
+        self.enabled = @NO;
+
+        // TODO: Remove this side effect
+        SentrySDK.logLevel = kSentryLogLevelError;
+        self.logLevel = kSentryLogLevelError;
+
+        self.debug = @NO;
+        self.maxBreadcrumbs = defaultMaxBreadcrumbs;
+        self.integrations = SentryOptions.defaultIntegrations;
+        self.sampleRate = @1;
+        self.enableAutoSessionTracking = @NO;
+        self.sessionTrackingIntervalMillis = [@30000 unsignedIntValue];
+        self.attachStacktrace = @NO;
+
+        // Set default release name
+        NSDictionary *infoDict = [[NSBundle mainBundle] infoDictionary];
+        if (nil != infoDict) {
+            self.releaseName =
+                [NSString stringWithFormat:@"%@@%@+%@", infoDict[@"CFBundleIdentifier"],
+                          infoDict[@"CFBundleShortVersionString"], infoDict[@"CFBundleVersion"]];
+        }
+    }
+    return self;
+}
+
 - (_Nullable instancetype)initWithDict:(NSDictionary<NSString *, id> *)options
                       didFailWithError:(NSError *_Nullable *_Nullable)error
 {
-    self = [super init];
-    if (self) {
+    if (self = [self init]) {
         [self validateOptions:options didFailWithError:error];
         if (nil != error && nil != *error) {
             [SentryLog
                 logWithMessage:[NSString stringWithFormat:@"Failed to initialize: %@", *error]
                       andLevel:kSentryLogLevelError];
             return nil;
         }
-
-        // If no user-defined release, use default.
-        if (nil == self.releaseName) {
-            NSDictionary *infoDict = [[NSBundle mainBundle] infoDictionary];
-            if (nil != infoDict) {
-                self.releaseName = [NSString
-                    stringWithFormat:@"%@@%@+%@", infoDict[@"CFBundleIdentifier"],
-                    infoDict[@"CFBundleShortVersionString"], infoDict[@"CFBundleVersion"]];
-            }
-        }
     }
     return self;
 }
 
+- (void)setDsn:(NSString *)dsn
+{
+    NSError *error = nil;
+    self.parsedDsn = [[SentryDsn alloc] initWithString:dsn didFailWithError:&error];
+    if (nil == error) {
+        _dsn = dsn;
+        self.enabled = @YES;
+    } else {
+        self.enabled = @NO;
+        NSString *errorMessage =
```

Comment:
M: can we throw an error instead? so they don't even need to monitor the logs, they will know something is wrong immediately

Reply:
@HazAT and I discussed this and our argument was that most of the time the user is going to copy-paste the dsn. We would have to do an `NSAssert` here, which is going to cause a crash if the dsn is not right. For us, that seems a little bit too strict. We can change it to `NSAssert`, but then we should also do this for `startWithOptions` and `initWithOptions` to be consistent IMO.

Reply:
I changed it to throw an `NSInvalidArgumentException`, but I didn't change it yet for `startWithOptions`. Instead I added it as a todo for 6.x.x.

Reply:
The downside throwing an Exception in Objective-C is that there is [no safe way to recover in Swift](https://developer.apple.com/documentation/swift/cocoa_design_patterns/handling_cocoa_errors_in_swift). @HazAT what do you think about this?

Reply:
I think it is fine like this, one of our rules is our SDK should never cause a crash/exception, so I prefer this over throwing something.

Reply:
Then let's just keep it as it was, because of what @HazAT stated and because there is no safe way to recover in Swift.